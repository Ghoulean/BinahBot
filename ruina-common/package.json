{
  "name": "@ghoulean/ruina-common",
  "private": true,
  "version": "1.0.0",
  "description": "",
  "main": "dist/index.js",
  "scripts": {
    "build": "tsc",
    "clean": "tsc --build --clean",
    "format": "npx eslint .",
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "author": "",
  "license": "ISC",
  "devDependencies": {
    "@typescript-eslint/eslint-plugin": "^5.60.0",
    "@typescript-eslint/parser": "^5.60.0",
    "eslint": "^8.43.0",
    "typescript": "^5.1.3"
  }
}
