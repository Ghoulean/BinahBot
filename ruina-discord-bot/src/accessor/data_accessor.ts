import {
    DecoratedAbnoPage,
    Localization,
    LookupResult,
} from "@ghoulean/ruina-common";
import * as fastestLevenshtein from "fastest-levenshtein";
import * as __AUTOCOMPLETE from "../data/autocomplete.json";
import * as __CN_ABNO from "../data/cn/abno.json";
import * as __EN_ABNO from "../data/en/abno.json";
import * as __JP_ABNO from "../data/jp/abno.json";
import * as __KR_ABNO from "../data/kr/abno.json";
import * as __LOOKUP_RESULTS from "../data/queryLookupResults.json";
import * as __TRCN_ABNO from "../data/trcn/abno.json";

type QueryToLookupResult = {
    [key: string]: LookupResult[];
};

type QueryToDecoratedAbnopage = {
    [key: string]: DecoratedAbnoPage;
};

type LocalizationToQueryDecoratedAbnopage = {
    [key in Localization]: QueryToDecoratedAbnopage;
};

type LevenshteinResults = {
    best: string;
    score: number;
};

const LOOKUP_RESULTS: QueryToLookupResult =
    __LOOKUP_RESULTS as QueryToLookupResult;

const CN_ABNO: QueryToDecoratedAbnopage = __CN_ABNO as QueryToDecoratedAbnopage;
const EN_ABNO: QueryToDecoratedAbnopage = __EN_ABNO as QueryToDecoratedAbnopage;
const JP_ABNO: QueryToDecoratedAbnopage = __JP_ABNO as QueryToDecoratedAbnopage;
const KR_ABNO: QueryToDecoratedAbnopage = __KR_ABNO as QueryToDecoratedAbnopage;
const TRCN_ABNO: QueryToDecoratedAbnopage =
    __TRCN_ABNO as QueryToDecoratedAbnopage;

const LOCALIZATION_TO_DECORATED_ABNO_PAGE: LocalizationToQueryDecoratedAbnopage =
    {
        [Localization.CHINESE_SIMPLIFIED]: CN_ABNO,
        [Localization.CHINESE_TRADITIONAL]: TRCN_ABNO,
        [Localization.ENGLISH]: EN_ABNO,
        [Localization.JAPANESE]: JP_ABNO,
        [Localization.KOREAN]: KR_ABNO,
    };

const AUTOCOMPLETE: string[] = __AUTOCOMPLETE.data;

const FUZZY_MATCHING_DISTANCE = 2;

export class DataAccessor {
    constructor() {}

    public lookup(query: string, preferredLocale: Localization): LookupResult {
        let lookupResults: LookupResult[] | undefined = LOOKUP_RESULTS[query];
        if (!lookupResults) {
            const levenshteinResults: LevenshteinResults =
                this.levenshteinLookup(query);
            if (levenshteinResults.score > FUZZY_MATCHING_DISTANCE) {
                throw new Error(`Couldn't identify query result for ${query}.`);
            }
            lookupResults = LOOKUP_RESULTS[levenshteinResults.best];
        }
        for (const lookupResult of lookupResults) {
            if (lookupResult.locale == preferredLocale) {
                return lookupResult;
            }
        }
        return lookupResults[0];
    }

    public getDecoratedAbnoPage(
        pageId: string,
        locale: Localization
    ): DecoratedAbnoPage {
        const decoratedAbnoPage: DecoratedAbnoPage | undefined =
            this.getLocaledAbnoMapping(locale)[pageId];
        if (!decoratedAbnoPage) {
            throw new Error(`${pageId} not found in ${locale} locale`);
        }
        return decoratedAbnoPage;
    }

    public autocomplete(query: string): string[] {
        const retVal: string[] = [];
        const cleanQuery = this.cleanQuery(query);
        for (const q of AUTOCOMPLETE) {
            if (
                this.autocompleteDistance(cleanQuery, q) <=
                FUZZY_MATCHING_DISTANCE
            ) {
                retVal.push(q);
            }
        }
        return retVal.sort((a, b) => {
            return (
                this.autocompleteDistance(cleanQuery, a) -
                    this.autocompleteDistance(cleanQuery, b) ||
                a.length - b.length ||
                a.localeCompare(b)
            );
        });
    }

    private getLocaledAbnoMapping(
        locale: Localization
    ): QueryToDecoratedAbnopage {
        return LOCALIZATION_TO_DECORATED_ABNO_PAGE[locale];
    }

    private cleanQuery(s: string): string {
        return s.replace(/\s/g, "").toLowerCase();
    }

    private autocompleteDistance(query: string, lookup: string): number {
        return fastestLevenshtein.distance(
            query,
            this.cleanQuery(lookup).substring(0, query.length)
        );
    }

    private levenshteinLookup(s: string): LevenshteinResults {
        const closest: string = fastestLevenshtein.closest(s, AUTOCOMPLETE);
        return {
            best: closest,
            score: fastestLevenshtein.distance(
                this.cleanQuery(s),
                this.cleanQuery(closest)
            ),
        };
    }
}
