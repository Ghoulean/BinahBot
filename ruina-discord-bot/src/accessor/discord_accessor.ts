import {
    GetSecretValueCommandOutput,
    SecretsManager,
} from "@aws-sdk/client-secrets-manager";
import axios from "axios";
import { EnvVarRetriever } from "../util/env_var_retriever";

const BASE_DISCORD_URL: string = "https://discord.com/api/v10";
// TODO: reconsider where this value goes
// TODO: rename
const DISCORD_SECRET_ENV_KEY: string = "DISCORD_BOT_API_KEY_NAME";

// TODO: reconsider where this type should go
type DiscordSecretsValues = {
    applicationId: string;
    authToken: string;
};

export class DiscordAccessor {
    private readonly secretsManager: SecretsManager;
    private readonly envVarRetriever: EnvVarRetriever;
    private applicationId: string;
    private botAuthToken: string;

    constructor(
        secretsManager: SecretsManager,
        envVarRetriever: EnvVarRetriever
    ) {
        this.secretsManager = secretsManager;
        this.envVarRetriever = envVarRetriever;
    }

    public async writeResponse(
        interactionToken: string,
        // TODO: strongly typed embed
        embed: any
    ): Promise<void> {
        console.log(`Retrieving secrets (if not done so already)`);
        await this.retrieveSecrets();
        console.log(`Got secrets`);
        try {
            const discordPayload: any = {
                embeds: [embed],
            };
            console.log(
                `DiscordAccessor: writing response with applicationId=${this.applicationId} interactionToken=${interactionToken}`
            );
            console.log(
                `DiscordAccessor: writing response with discordPayload=${JSON.stringify(
                    discordPayload
                )}`
            );
            const { status } = await axios.post(
                `${BASE_DISCORD_URL}/webhooks/${this.applicationId}/${interactionToken}`,
                discordPayload,
                {
                    headers: {
                        Authorization: `Bot ${this.botAuthToken}`,
                        "Content-Type": "application/json",
                    },
                }
            );
            if (status === 204) {
                console.log("Successfully returned Discord response");
            }
        } catch (e) {
            throw e;
        }
    }

    private async retrieveSecrets(): Promise<void> {
        if (this.botAuthToken) {
            return;
        }
        const getSecretValueCommandOutput: GetSecretValueCommandOutput =
            await this.secretsManager.getSecretValue({
                SecretId: this.envVarRetriever.getRequired(
                    DISCORD_SECRET_ENV_KEY
                ),
            });
        if (getSecretValueCommandOutput.SecretString) {
            const secrets: DiscordSecretsValues = JSON.parse(
                getSecretValueCommandOutput.SecretString
            ) as DiscordSecretsValues;
            this.applicationId = secrets.applicationId;
            this.botAuthToken = secrets.authToken;
        }
    }
}
