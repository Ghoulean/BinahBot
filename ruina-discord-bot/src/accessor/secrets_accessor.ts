import {
    GetSecretValueCommand,
    GetSecretValueCommandInput,
    GetSecretValueCommandOutput,
    SecretsManager,
} from "@aws-sdk/client-secrets-manager";
import { DiscordSecrets } from "../model/discord/discord_secrets";

export class SecretsAccessor {
    private readonly secretsId: string;
    private readonly secretsManager: SecretsManager;
    private discordSecrets: DiscordSecrets;

    constructor(secretsId: string, secretsManager: SecretsManager) {
        this.secretsId = secretsId;
        this.secretsManager = secretsManager;
    }

    public async getDiscordSecrets(): Promise<DiscordSecrets> {
        return this.discordSecrets ?? await this.cacheDiscordSecrets();
    }

    private async cacheDiscordSecrets(): Promise<DiscordSecrets> {
        const input: GetSecretValueCommandInput = {
            SecretId: this.secretsId,
        };
        const command: GetSecretValueCommand = new GetSecretValueCommand(input);
        let discordApiKeys: GetSecretValueCommandOutput;
        try {
            discordApiKeys = await this.secretsManager.send(command);
        } catch (e: unknown) {
            throw new Error(`Unable to get Discord secrets: ${JSON.stringify(e)}`);
        }
        if (discordApiKeys.SecretString) {
            this.discordSecrets = JSON.parse(
                discordApiKeys.SecretString
            ) as DiscordSecrets;
        } else {
            throw new Error(
                `Got Discord secrets, but can't parse SecretString in ${JSON.stringify(discordApiKeys)}`
            );
        }
        return this.discordSecrets;
    }
}
