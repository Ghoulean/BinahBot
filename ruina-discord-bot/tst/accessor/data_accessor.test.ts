import { Localization } from "@ghoulean/ruina-common";
import { DataAccessor } from "../../src/accessor/data_accessor";

const WEIGHT_OF_SIN_EN_PAGE_NAME = "The Weight of Sin";
const WEIGHT_OF_SIN_PAGE_ID = "LongBird_Sin";
const BAD_PAGE_NAME = "PageNameThatDoesNotExist";
const BAD_PAGE_ID = "PageIdThatDoesNotExist";

let dataAccessor: DataAccessor;

beforeEach(() => {
    dataAccessor = new DataAccessor();
});

test("should return lookup result when given page name and locale", () => {
    expect(
        dataAccessor.lookup(WEIGHT_OF_SIN_EN_PAGE_NAME, Localization.ENGLISH)
    ).toMatchSnapshot();
});

test("should throw error when given bad lookup request", () => {
    expect(() => {
        dataAccessor.lookup(BAD_PAGE_NAME, Localization.ENGLISH);
    }).toThrow();
});

test("should return decorated abno page result when given page id and locale", () => {
    expect(
        dataAccessor.getDecoratedAbnoPage(
            WEIGHT_OF_SIN_PAGE_ID,
            Localization.ENGLISH
        )
    ).toMatchSnapshot();
});

test("should throw error when given bad decorated abno page request", () => {
    expect(() => {
        dataAccessor.getDecoratedAbnoPage(BAD_PAGE_ID, Localization.ENGLISH);
    }).toThrow();
});
