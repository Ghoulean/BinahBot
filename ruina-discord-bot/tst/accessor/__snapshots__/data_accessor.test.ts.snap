// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return autocomplete result when given query 1`] = `
[
  "The Beast",
  "The Finale",
  "The Fairies' Care",
  "The Weight of Sin",
  "The Seventh Bullet",
  "The Thirteenth Toll",
  "The Role of the Wolf",
]
`;

exports[`should return decorated abno page result when given page id and locale 1`] = `
{
  "chapter": 6,
  "description": "[Single Ally]
Whenever a die this librarian plays rolls the minimum value, take 2-7 damage and gain 1-2 Strength and Endurance next Scene.",
  "emoLevel": 1,
  "emotionRate": 0,
  "emotionSign": -1,
  "flavorText": "Its scale would never neglect the weight of even the smallest sin.",
  "floor": "Binah",
  "imagePath": "/LongBird_Sin.png",
  "locale": "en",
  "name": "The Weight of Sin",
  "nameId": "LongBird_Sin",
  "scriptId": "longbird1",
  "targetType": "SelectOne",
}
`;

exports[`should return lookup result when given page name and locale 1`] = `
{
  "chapter": 6,
  "locale": "en",
  "pageId": "LongBird_Sin",
  "pageType": 0,
  "query": "The Weight of Sin",
}
`;
